src/main/java/org/springframework/data/convert/PropertyValueConverterFactories.java

```diff
@@ -0,0 +1,240 @@
+							AutowireCapableBeanFactory.AUTOWIRE_CONSTRUCTOR, false);
+				}
+			}
+			return null;
+		}
+	}
+
+	/**
+	 * {@link PropertyValueConverterFactory} implementation that serves {@link PropertyValueConverter} from a given
+	 * {@link ValueConverterRegistry registry}.
+	 * 
+	 * @author Christoph Strobl
+	 * @since 2.7
+	 */
+	static class ConfiguredInstanceServingValueConverterFactory implements PropertyValueConverterFactory {
+
+		private final ValueConverterRegistry<?> converterRegistry;
+
+		ConfiguredInstanceServingValueConverterFactory(ValueConverterRegistry<?> converterRegistry) {
+
+			Assert.notNull(converterRegistry, "ConversionsRegistrar must not be null!");
+			this.converterRegistry = converterRegistry;
+		}
+
+		@Nullable
+		@Override
+		public <A, B, C extends ValueConversionContext<?>> PropertyValueConverter<A, B, C> getConverter(
+				PersistentProperty<?> property) {
+			return (PropertyValueConverter<A, B, C>) converterRegistry.getConverter(property.getOwner().getType(),
+					property.getName());
+		}
+
+		@Override
+		public <S, T, C extends ValueConversionContext<?>> PropertyValueConverter<S, T, C> getConverter(
+				Class<? extends PropertyValueConverter<S, T, C>> converterType) {
+			return null;
+		}
+	}
+
+	/**
+	 * {@link PropertyValueConverterFactory} implementation that caches converters provided by an underlying
+	 * {@link PropertyValueConverterFactory factory}.
+	 * 
+	 * @author Christoph Strobl
+	 * @since 2.7
+	 */
+	static class CachingPropertyValueConverterFactory implements PropertyValueConverterFactory {
+
+		private final PropertyValueConverterFactory delegate;
+		private final Cache cache = new Cache();
+
+		CachingPropertyValueConverterFactory(PropertyValueConverterFactory delegate) {
+
+			Assert.notNull(delegate, "Delegate must not be null!");
+			this.delegate = delegate;
+		}
+
+		@Nullable
+		@Override
+		public <S, T, C extends ValueConversionContext<?>> PropertyValueConverter<S, T, C> getConverter(
+				PersistentProperty<?> property) {
+
+			Optional<PropertyValueConverter<?, ?, ? extends ValueConversionContext<?>>> converter = cache.get(property);
+
+			return converter != null ? (PropertyValueConverter) converter.orElse(null)
+					: cache.cache(property, delegate.getConverter(property));
+		}
+
+		@Override
+		public <S, T, C extends ValueConversionContext<?>> PropertyValueConverter<S, T, C> getConverter(
+				Class<? extends PropertyValueConverter<S, T, C>> converterType) {
+
+			Optional<PropertyValueConverter<?, ?, ? extends ValueConversionContext<?>>> converter = cache.get(converterType);
+
+			return converter != null ? (PropertyValueConverter) converter.orElse(null)
+					: cache.cache(converterType, delegate.getConverter(converterType));
+		}
+
+		static class Cache {
+
+			Map<PersistentProperty<?>, Optional<PropertyValueConverter<?, ?, ? extends ValueConversionContext<?>>>> perPropertyCache = new HashMap<>();
```

Comment:
We either need synchronization or should use `ConcurrentHashMap` to avoid concurrent modification exceptions.